jazkarta/shop/browser/checkout/__init__.py

```diff
@@ -0,0 +1,202 @@
+from datetime import date
+from plone.protect.utils import safeWrite
+from premailer import Premailer
+from Products.Five import BrowserView
+from zope.browserpage import ViewPageTemplateFile
+from zope.cachedescriptors.property import Lazy as lazy_property
+from zope.event import notify
+from zope.interface import implementer
+from ...cart import Cart
+from ...interfaces import CheckoutComplete
+from ...interfaces import IDontShowJazkartaShopPortlets
+from ...interfaces import TaxRateException
+from ...utils import get_setting
+from ...utils import has_permission
+from ...utils import run_in_transaction
+from ...utils import send_mail
+from ...utils import PLONE_VERSION
+from ...vocabs import country_names
+
+
+class P5Mixin():
+    """ utility method to distinguish between Plone 4 and 5 """
+
+    def using_plone5(self):
+        if PLONE_VERSION[0] == '5':
+            return True
+        return False
+
+
+@implementer(IDontShowJazkartaShopPortlets)
+class CheckoutForm(BrowserView):
+
+    def __call__(self):
+        from .authorize_net_accept_js import CheckoutFormAuthorizeNetAcceptJs
+        from .authorize_net_sim import CheckoutFormAuthorizeNetSIM
+        from .stripe import CheckoutFormStripe
+
+        payment_processor = get_setting('payment_processor')
+        if payment_processor == 'Authorize.Net SIM':
+            return CheckoutFormAuthorizeNetSIM(self.context, self.request)()
+        elif payment_processor == 'Authorize.Net Accept.js':
+            return CheckoutFormAuthorizeNetAcceptJs(
+                self.context, self.request)()
+        elif payment_processor == 'Stripe':
+            return CheckoutFormStripe(self.context, self.request)()
+        else:
+            raise Exception(
+                    'No valid payment processor has been specified.')
+
+
+class CheckoutFormBase(P5Mixin):
```

Comment:
Wait. Shouldn't this subclass BrowserView as well? Otherwise, it can't take context and request as constructor arguments, can it?

Reply:
Hmm, I think the zcml directive that registers the checkout view classes generates a new class that mixes in BrowserView. You're right, it would probably be better to be explicit, though I don't think it'll change the functionality.

Reply:
It might be that it does that in Plone 5, but not Plone 4. I'm getting a "this constructor takes no arguments" error now that I'm trying this out.

Reply:
Ah, I think the difference is that on NCI I overrode @@checkout to point directly at a customized checkout form, so it's getting the ZCML-registered version with the automatic mixin. Here the forms are referenced directly in Python (from `CheckoutForm`) so it doesn't find a ZCML-registered view. I think you're right that we should add BrowserView as a base class here.